client/signup/steps/page-picker/use-cart-for-difm.tsx

```diff
@@ -0,0 +1,248 @@
+import { getCurrentPlan } from 'calypso/state/sites/plans/selectors';
+import { isCurrentPlanPaid } from 'calypso/state/sites/selectors';
+import type {
+	MinimalRequestCartProduct,
+	ResponseCart,
+	ResponseCartProduct,
+} from '@automattic/shopping-cart';
+import type { ProductListItem } from 'calypso/state/products-list/selectors/get-products-list';
+import type { TranslateResult } from 'i18n-calypso';
+
+export type CartItem = {
+	nameOverride?: TranslateResult;
+	product:
+		| ( Partial< ProductListItem > & Pick< ProductListItem, 'cost' | 'product_name' > )
+		| ( Partial< ResponseCartProduct > & Pick< ResponseCartProduct, 'cost' | 'product_name' > );
+	meta?: TranslateResult;
+	productCount?: number;
+	lineCost: number;
+};
+
+const FREE_PAGES = 5;
+
+type DummyCartParams = {
+	selectedPages: string[];
+	currencyCode: string;
+	activePlanScheme: ProductListItem | null;
+	difmLiteProduct: ProductListItem | null;
+	extraPageProduct: ProductListItem | null;
+	translate: LocalizeProps[ 'translate' ];
+	formatCurrency: ( original: number, currrency: string ) => string | null;
+};
+
+function getDummyCartProducts( {
+	selectedPages,
+	currencyCode,
+	activePlanScheme,
+	difmLiteProduct,
+	extraPageProduct,
+	translate,
+	formatCurrency,
+}: DummyCartParams ): CartItem[] {
+	const extraPageCount = Math.max( 0, selectedPages.length - FREE_PAGES );
+	let displayedCartItems: CartItem[] = [];
+	if ( difmLiteProduct && activePlanScheme && extraPageProduct ) {
+		displayedCartItems = [
+			{
+				nameOverride: translate( 'Website Design Service' ),
+				product: difmLiteProduct,
+				lineCost: difmLiteProduct.cost,
+				meta: translate( 'One-time fee' ),
+			},
+			{
+				product: activePlanScheme,
+				meta: translate( 'Plan Subscription: %(planPrice)s per year', {
+					args: { planPrice: formatCurrency( activePlanScheme.cost, currencyCode ) },
+				} ),
+				lineCost: activePlanScheme.cost,
+			},
+
+			{
+				nameOverride: `${ extraPageCount } ${
+					extraPageCount === 1 ? translate( 'Extra Page' ) : translate( 'Extra Pages' )
+				}`,
+				product: extraPageProduct,
+				meta: translate( '%(perPageCost)s Per Page', {
+					args: {
+						perPageCost: extraPageProduct.cost_display,
+					},
+				} ),
+				lineCost: extraPageProduct.cost * extraPageCount,
+				productCount: extraPageCount,
+			},
+		];
+	}
+
+	// Hide pro plan if existing site has a paid plan
+	if ( activePlanScheme ) {
+		displayedCartItems = displayedCartItems.filter(
+			( p ) => p.product.product_slug !== activePlanScheme.product_slug
+		);
+	}
+	return displayedCartItems;
+}
+
+function getSiteCartProducts( {
+	responseCart,
+	translate,
+}: {
+	responseCart: ResponseCart;
+	translate: LocalizeProps[ 'translate' ];
+} ): CartItem[] {
+	return responseCart.products.map( ( product ) => {
+		switch ( product.product_slug ) {
+			case PLAN_WPCOM_PRO:
+				return {
+					product: product,
+					lineCost: product.cost,
```

Comment:
While there's nothing broken with this approach, I wanted to mention something important to know about dealing with prices.

`cost` is a deprecated property because it is a float. We typically use `item_subtotal_integer` (which is in the currency's smallest unit if we need to do math on the number - math on floating point numbers is unreliable on computers and we want to be accurate for prices) or `item_subtotal_display` (which is a string with decimals and a currency symbol if we need to display the localized cost). (`cost` also doesn't include `volume`.)

It looks like below you're doing math on these numbers, so I'd recommend using `item_subtotal_integer`, but then you'll need to convert them again to be able to format them into a localized format. This is hard to do well, which is why we typically leave all pricing math to backend operations. 

Are the calculations you're doing here something that could be done on the backend in an endpoint response?

Reply:
Thanks for raising these @sirbrillig , these are valid concern. However, it seems this PR is now on hold pending a CFT and it most probably might be scrapped. However if we decide to go through with this. I will  Look into coming up with an endpoint to calculate the total!

Reply:
No problem. And if you do decide to use this PR, and the cost of making new endpoints is prohibitive, you can probably still use the approach you have here. We may be able to return more useful data from the shopping cart endpoint to help. Just let me know! Dealing with prices is a pain. 😁